### PR TITLE
scheduler/balance_leader: fix data race in the function of clone for config

### LINKED ISSUE #5156: scheduler/balance_leader: potential data race in the function of clone for config
## Bug Report

It may cause data race since it do not deeply clone the slice of `conf.Ranges` in the function of `Clone` for config:
[https://github.com/tikv/pd/blob/c8775b6176a03eb25cb1622fc46ba1d496872df7/server/schedulers/balance_leader.go#L128-L133](https://github.com/tikv/pd/blob/c8775b6176a03eb25cb1622fc46ba1d496872df7/server/schedulers/balance_leader.go#L128-L133)

### PR DESCRIPTION
**Pull request by AndreMouche**
Signed-off-by: shirly <AndreMouche@126.com>

### What problem does this PR solve?

Issue Number: Close #5156 

### Check List

Tests

- Unit test

### Release note

```release-note
 None.
```

### CODE REVIEW

**Review by CabinfeverB — APPROVED** [review]

File: server/schedulers/balance_leader.go [hunk 1]
```diff
@@ -128,8 +128,9 @@ func (conf *balanceLeaderSchedulerConfig) validate() bool {
 func (conf *balanceLeaderSchedulerConfig) Clone() *balanceLeaderSchedulerConfig {
 	conf.mu.RLock()
 	defer conf.mu.RUnlock()
+	var ranges []core.KeyRange
```
Comment (bufferflies) [inline server/schedulers/balance_leader.go:131]: How about define the capacity for the ranges first?  some difference between copy and append : https://stackoverflow.com/questions/71316774/is-it-equivalent-use-copy-or-append-slice-type-in-golang

File: server/schedulers/balance_leader_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+package schedulers
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestBalanceLeaderSchedulerConfigClone(t *testing.T) {
```
Comment (bufferflies) [inline server/schedulers/balance_leader_test.go:23]: Maybe not need this unit test?

Comment (AndreMouche) [reply server/schedulers/balance_leader_test.go:23]: The test is used to prove this PR which should be failed without this PR.

**Comment by rleungx** [comment]
Does other scheduler have the same problems?

**Comment by AndreMouche** [comment]
> Does other scheduler have the same problems?

Yes, I've review schedulers, and find that almost all Clone function for those config that contains slice or map, exist the similar problem.
How about create one or more issues and fix it one by one?

**Comment by rleungx** [comment]
Does this PR need to cherry-pick? If so, I suggest merging them into one.

**Comment by AndreMouche** [comment]
Maybe not? since the read function happens rarely

> Does this PR need to cherry-pick? If so, I suggest merging them into one.

File: server/schedulers/balance_leader_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	re := require.New(t)
+	keyRanges1, _ := getKeyRanges([]string{"a", "b", "c", "d"})
+	conf := &balanceLeaderSchedulerConfig{
+		Ranges: keyRanges1,
+		Batch:  10,
+	}
+	con2 := conf.Clone()
```
Comment (rleungx) [inline server/schedulers/balance_leader_test.go:30]:
```suggestion
	conf2 := conf.Clone()
```

**Review by rleungx — APPROVED** [review]

**Pull request merged**